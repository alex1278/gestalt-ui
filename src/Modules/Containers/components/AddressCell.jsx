import React, { memo } from 'react';
import PropTypes from 'prop-types';
import styled from 'styled-components';
import { A } from 'components/Links';
import { ClipboardButton } from 'components/Buttons';

const Cell = styled.div`
  display: inline-flex;
  flex: 1;
  align-items: center;
`;

const AddressCell = memo(({ address, copyClip }) => {
  const href = `${address.protocol}://${address.host}:${address.port}`;

  if (Object.keys(address).length > 0 && !href.includes('undefined')) {
    return (
      <Cell>
        {copyClip &&
          <ClipboardButton
            showLabel={false}
            text={href}
            tooltipLabel="Copy to clipboard"
          />}
        <A
          href={href}
          target="_blank"
          rel="noopener noreferrer"
          primary
        >
          {href}
        </A>
      </Cell>
    );
  }

  return null;
});

AddressCell.propTypes = {
  address: PropTypes.shape({
    protocol: PropTypes.string,
    host: PropTypes.string,
    port: PropTypes.number,
  }),
  copyClip: PropTypes.bool
};

AddressCell.defaultProps = {
  address: {},
  copyClip: false,
};

export default AddressCell;
