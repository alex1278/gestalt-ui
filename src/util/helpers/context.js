/**
 * generateContextEntityState
 * @param {*} params
 */
export function generateContextEntityState(params) {
  const entity = {
    id: params.environmentId || params.workspaceId || null,
    key: null,
  };

  if (params.workspaceId && params.environmentId) {
    entity.key = 'environments';
  }

  if (params.workspaceId && !params.environmentId) {
    entity.key = 'workspaces';
  }

  // if (params.workspaceId && params.environmentId && params.providerId) {
  //   entity.id = params.providerId;
  //   entity.key = 'providers';
  // }

  return entity;
}
