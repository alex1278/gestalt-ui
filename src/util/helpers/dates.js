import moment from 'moment';

export const getNow = () => moment();

export const getAgo = (num = 0, unit = 'day') =>
  moment().subtract(num, unit);

export const isRecentWithin = (timestamp, num, unit) =>
  moment(timestamp).isBetween(getAgo(num, unit), getNow());
