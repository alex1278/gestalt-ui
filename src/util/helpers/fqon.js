// sad hack for parsing url

export default function fqonParent(organization) {
  if (organization.org) {
    const urlParser = document.createElement('a');
    urlParser.href = organization.org.href;
    return urlParser.pathname.replace('/', '').split('/')[0];
  }

  return organization;
}
