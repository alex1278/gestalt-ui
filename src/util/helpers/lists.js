export function insertItem(array, payload) {
  return [...array, payload];
}

export function removeItem(array, payload) {
  const newArray = array.slice();
  newArray.splice(newArray.findIndex(a => a === payload), 1);

  return newArray;
}

export function removeItemById(array, id, keyField = 'id') {
  const newArray = array.slice();
  newArray.splice(newArray.findIndex(a => a[keyField] === id), 1);

  return newArray;
}

export function updateItem(array, payload) {
  return array.map((item) => {
    if (item.id === payload.id) {
      return {
        ...payload,
      };
    }

    // Leave every other item unchanged
    return item;
  });
}
