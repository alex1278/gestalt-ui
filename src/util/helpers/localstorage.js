/**
 * getItem
 * @param {*} key
 */
export const getItem = (key) => {
  try {
    return localStorage.getItem(key);
  } catch (err) {
    return undefined;
  }
};

/**
 * saveItem
 * @param {*} key
 * @param {*} value
 */
export const saveItem = (key, value) => {
  try {
    localStorage.setItem(key, value);
  } catch (err) {
    // ignore write errors
  }

  return undefined;
};

/**
 * removeItem
 * @param {*} key
 * @param {*} value
 */
export const removeItem = (key) => {
  try {
    localStorage.removeItem(key);
  } catch (err) {
    // ignore write errors
  }

  return undefined;
};
