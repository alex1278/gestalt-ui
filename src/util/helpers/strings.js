import { startCase, toLower } from 'lodash';

/**
 * getLastFromSplit
 * Parses and returns the last item in a structure like GrandPareng::Parent::Child
 * @param {String} string
 */
export function getLastFromSplit(string, delimiter = '::') {
  if (string) {
    const split = string.split(delimiter);

    return split[split.length - 1];
  }
  return string;
}

/**
 * truncate
 * @param {String} string
 * @param {Number} at
 * @param {String} trailing
 */
export function truncate(string, at, trailing = '...') {
  if (string) {
    if (string.length < at) {
      return string;
    }

    return `${string.substring(0, at).trim()}${trailing}`;
  }

  return string;
}

/**
 * toTitleCase
 * @param {String} string
 */
export function toTitleCase(string) {
  return startCase(toLower(string));
}


// TODO: Will be rmeoved when we have security config
export const checkIfPassword = field =>
  field && field.length && (field.toUpperCase().includes('PASSWORD') || field.toUpperCase().includes('SECRET') || field.toUpperCase().includes('KEY'));

export default {
  getLastFromSplit,
  truncate,
  toTitleCase,
  checkIfPassword
};

export const removeHostFromURL = (url) => {
  if (url) {
    return url.replace(/^.*\/\/[^/]+:?[0-9]?\//i, '');
  }

  return url;
};

export const isBase64 = (str) => {
  try {
    return btoa(atob(str)) === str;
  } catch (err) {
    return false;
  }
};
