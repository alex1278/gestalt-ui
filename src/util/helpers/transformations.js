import map from 'lodash/map';
import compact from 'lodash/compact';
import get from 'lodash/get';

/**
 * arrayToMap
 * @param {*} array
 * @param {*} keyName
 * @param {*} valueName
 */
export function arrayToMap(array = [], keyName = 'name', valueName = 'value') {
  if (!Array.isArray(array)) {
    return array;
  }

  return Object.assign({}, ...array.map(v => ({ [v[keyName]]: v[valueName] || '' })));
}

/**
 * mapTo2DArray
 * @param {*} object
 * @param {*} keyName
 * @param {*} valueName
 * @param {*} mergeSet
 */
export function mapTo2DArray(object = {}, keyName = 'name', valueName = 'value', mergeSet = {}) {
  if (typeof object !== 'object') {
    return object;
  }

  return map(object, (value, key) => (Object.assign({ [keyName]: key, [valueName]: value || '' }, mergeSet)));
}

/**
 * stringDemiltedToArray
 * convert comma delimited string to an array and remove blank entries
 * @param {*} string
 */
export function stringDemiltedToArray(string) {
  return compact(string.split(','));
}

/**
 * convertFromMaps
 * used to convert env vars from a map to an array and reconcile inherited vs own
 * @param {*} own
 * @param {*} inherited
 */
export function convertFromMaps(own = {}, inherited = {}, keyName = 'name', valueName = 'value') {
  const ownVars = mapTo2DArray(own, keyName, valueName, { inherited: false });

  // base case just return the mapped own vars
  if (!Object.keys(inherited)) {
    return ownVars;
  }

  const inheritedVars = mapTo2DArray(inherited, keyName, valueName, { inherited: true, overridden: false });

  const reconcileInheritedVars = inheritedVars.map((i) => {
    // find the dupe value in ownVars
    const index = ownVars.findIndex(o => o[keyName] === i[keyName]);

    // if the value is differnt mark the matching var as overriden
    if (index > -1 && ownVars[index][valueName] !== i[valueName]) {
      return { ...i, [valueName]: ownVars[index][valueName], inherited: true, overridden: true };
    }

    // default case the own env var
    return i;
  });

  const results = [...reconcileInheritedVars];

  // "merge" the ownVars back in
  ownVars.forEach((v) => {
    const index = inheritedVars.findIndex(o => o[keyName] === v[keyName]);

    if (index === -1) {
      results.push(v);
    }
  });

  return results;
}

/**
 * Builds an object from a delimited string
 * @param {String} string
 * @param {String} delimiter
 */
export function nestedObjectFromString(string, value = null, delimeter = '.') {
  const [last, ...paths] = string.split(delimeter).reverse();
  return paths.reduce((acc, el) => ({ [el]: acc }), { [last]: value });
}


export const getRoutePattern = (object, urlPattern) => {
  const keys = [];

  urlPattern.split('/').forEach((item) => {
    const search = item.startsWith(':');

    if (search) {
      keys.push(get(object, item.slice(1)));
    } else {
      keys.push(item);
    }
  });

  return keys.join('/');
};
