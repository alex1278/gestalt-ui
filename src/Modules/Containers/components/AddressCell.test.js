import React from 'react';
import AddressCell from './AddressCell';
import { renderWithTheme } from '../../../../test/helpers';

describe('Containers::Component AddressCell', () => {
  it('does not render when there is no address', () => {
    const { container } = renderWithTheme(<AddressCell address={{}} />);

    expect(container.firstChild).toBeNull();
  });

  it('does not render when there is an invalid address shape', () => {
    const address = { host: 'gazoropazorp', port: 80, };
    const { container } = renderWithTheme(<AddressCell address={address} />);

    expect(container.firstChild).toMatchSnapshot();
  });

  it('renders with a valid address', () => {
    const address = { protocol: 'http', host: 'gazoropazorp', port: 80, };
    const { container } = renderWithTheme(<AddressCell address={address} />);

    expect(container.firstChild).toMatchSnapshot();
  });
});
