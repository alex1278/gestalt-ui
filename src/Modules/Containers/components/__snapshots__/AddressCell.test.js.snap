// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Containers::Component AddressCell does not render when there is an invalid address shape 1`] = `null`;

exports[`Containers::Component AddressCell renders with a valid address 1`] = `
.c1 {
  font-size: 13px;
  -webkit-text-decoration: none;
  text-decoration: none;
  color: #03a9f4;
}

.c0 {
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  -webkit-flex: 1;
  -ms-flex: 1;
  flex: 1;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
}

<div
  class="c0"
>
  <a
    class="c1"
    href="http://gazoropazorp:80"
    rel="noopener noreferrer"
    target="_blank"
  >
    http://gazoropazorp:80
  </a>
</div>
`;
